cpp/tests/mtmg/multi_node_threaded_test.cu

```diff
@@ -0,0 +1,603 @@
+#include <cugraph/mtmg/vertex_result.hpp>
+
+#include <raft/util/cudart_utils.hpp>
+
+#include <rmm/device_uvector.hpp>
+
+#include <gtest/gtest.h>
+
+#include <nccl.h>
+
+#include <filesystem>
+#include <fstream>
+#include <vector>
+
+#include <thrust/count.h>
+#include <thrust/unique.h>
+
+struct Multithreaded_Usecase {
+  bool test_weighted{false};
+  bool check_correctness{true};
+};
+
+std::string g_comms_dir_name{};
+int g_node_rank{};
+int g_num_nodes{};
+int g_execution_id{0};
+
+template <typename input_usecase_t>
+class Tests_Multithreaded
+  : public ::testing::TestWithParam<std::tuple<Multithreaded_Usecase, input_usecase_t>> {
+ public:
+  Tests_Multithreaded() {}
+
+  static void SetUpTestCase() {}
+  static void TearDownTestCase() {}
+
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  std::vector<int> get_gpu_list()
+  {
+    int num_gpus_per_node{1};
+    RAFT_CUDA_TRY(cudaGetDeviceCount(&num_gpus_per_node));
+
+    std::vector<int> gpu_list(num_gpus_per_node);
+    std::iota(gpu_list.begin(), gpu_list.end(), 0);
+
+    return gpu_list;
+  }
+
+  void wait_for_directory(std::string directory_name, int max_tries = 60)
+  {
+    while (max_tries > 0) {
+      if (std::filesystem::is_directory(directory_name)) break;
+      sleep(1);
+      --max_tries;
+    }
+
+    CUGRAPH_EXPECTS(std::filesystem::is_directory(directory_name),
+                    "Timed out waiting for directory to be created");
+  }
+
+  std::ifstream wait_for_file(std::string file_name, int max_tries = 60)
+  {
+    while (max_tries > 0) {
+      if (std::filesystem::is_regular_file(file_name)) break;
+      sleep(1);
+      --max_tries;
+    }
+
+    CUGRAPH_EXPECTS(std::filesystem::is_regular_file(file_name),
+                    "Timed out waiting for file to be created");
+
+    return std::ifstream(file_name, std::ios::binary);
+  }
+
+  template <typename vertex_t,
+            typename edge_t,
+            typename weight_t,
+            typename result_t,
+            bool multi_gpu>
+  void run_current_test(
+    std::tuple<Multithreaded_Usecase const&, input_usecase_t const&> const& param,
+    std::vector<int> gpu_list)
+  {
+    using edge_type_t = int32_t;
+
+    constexpr bool renumber           = true;
+    constexpr bool do_expensive_check = false;
+
+    auto [multithreaded_usecase, input_usecase] = param;
+
+    raft::handle_t handle{};
+
+    result_t constexpr alpha{0.85};
+    result_t constexpr epsilon{1e-6};
+
+    size_t device_buffer_size{64 * 1024 * 1024};
+    size_t thread_buffer_size{4 * 1024 * 1024};
+
+    int num_local_gpus = gpu_list.size();
+    int num_threads    = num_local_gpus * 4;
+
+    //
+    //  This is intended to mimic a multi-node host application (non-MPI) integrating
+    //  with MTMG library.  This is a simple implementation using a shared file system
+    //  to pass configuration messages.  Terribly inefficient, but should mimic
+    //  expected behavior.
+    //
```

Comment:
What do you think about this vs using mpi but creating just one process per node? We are already using MPI for testing, so the latter won't require a new execution environment. The former assumes availability of a sort of NFS and users need to setup the directory path correctly. This is not necessary if we just re-use the existing MPI test environment but just create one process per node.

Reply:
I was think more like updating

```
#define CUGRAPH_MG_TEST_PROGRAM_MAIN()                                                  \
  int main(int argc, char** argv)                                                       \
  {                                                                                     \
    cugraph::test::initialize_mpi(argc, argv);                                          \
    auto comm_rank = cugraph::test::query_mpi_comm_world_rank();                        \
    auto comm_size = cugraph::test::query_mpi_comm_world_size();                        \
    int num_gpus_per_node{};                                                            \
    RAFT_CUDA_TRY(cudaGetDeviceCount(&num_gpus_per_node));                              \
    RAFT_CUDA_TRY(cudaSetDevice(comm_rank % num_gpus_per_node));                        \
    ::testing::InitGoogleTest(&argc, argv);                                             \
    auto const cmd_opts = parse_test_options(argc, argv);                               \
    auto const rmm_mode = cmd_opts["rmm_mode"].as<std::string>();                       \
    auto resource       = cugraph::test::create_memory_resource(rmm_mode);              \
    rmm::mr::set_current_device_resource(resource.get());                               \
    cugraph::test::g_perf = cmd_opts["perf"].as<bool>();                                \
    cugraph::test::g_rmat_scale =                                                       \
      (cmd_opts.count("rmat_scale") > 0)                                                \
        ? std::make_optional<size_t>(cmd_opts["rmat_scale"].as<size_t>())               \
        : std::nullopt;                                                                 \
    cugraph::test::g_rmat_edge_factor =                                                 \
      (cmd_opts.count("rmat_edge_factor") > 0)                                          \
        ? std::make_optional<size_t>(cmd_opts["rmat_edge_factor"].as<size_t>())         \
        : std::nullopt;                                                                 \
    cugraph::test::g_test_file_name =                                                   \
      (cmd_opts.count("test_file_name") > 0)                                            \
        ? std::make_optional<std::string>(cmd_opts["test_file_name"].as<std::string>()) \
        : std::nullopt;                                                                 \
                                                                                        \
    auto ret = RUN_ALL_TESTS();                                                         \
    cugraph::test::finalize_mpi();                                                      \
    return ret;                                                                         \
  }
```

and create  `CUGRAPH_MTMG_TEST_PROGRAM_MAIN()` We can just broadcast NCCL unique ID here and we won't touch anything related to MPI after this (and we can even call cugraph::test::finalize_mpi() right after broadcasting NCCL unique ID).

Reply:
Got the MPI configuration working in the next push.  I did not abstract it into a reusable function/MACRO, I can look at that when we added a second test (perhaps in a PR for adding Louvain or Jaccard).